File: src/cli-bridge/parser.ts
import type { CliBridgeAgent, CliHookMessage, CliHookMessageType, CliQuestionItem } from './types';

function normalizeEventName(eventName?: string): string | undefined {
  if (!eventName) return eventName;
  if (['PermissionRequest', 'permission_request', 'permission.asked', 'permission_requested'].includes(eventName)) return 'PermissionRequest';
  if (['PreToolUse', 'pre_tool_use'].includes(eventName)) return 'PreToolUse';
  if (['PostToolUse', 'post_tool_use'].includes(eventName)) return 'PostToolUse';
  if (['Stop', 'stop', 'SubagentStop', 'subagent_stop', 'session.idle', 'session_idle'].includes(eventName)) return 'TaskComplete';
  if (['StopFailure', 'stop_failure', 'session.error', 'session_error'].includes(eventName)) return 'TaskCompleteFailure';
  return eventName;
}

function stringifySummaryValue(value: unknown): string {
  if (value == null) return '';
  if (typeof value === 'string') return value.trim();
  if (typeof value === 'number' || typeof value === 'boolean') return String(value);
  if (Array.isArray(value)) return value.map(stringifySummaryValue).filter(Boolean).join('\n').trim();
  if (typeof value === 'object') {
    const obj = value as Record<string, unknown>;
    for (const key of ['last_assistant_message', 'lastAssistantMessage', 'assistant_message', 'assistantMessage', 'assistant', 'final', 'completion', 'answer', 'response', 'output', 'result', 'content', 'text', 'message', 'summary', 'error']) {
      const text = stringifySummaryValue(obj[key]);
      if (text) return text;
    }
  }
  return '';
}

export function parseHookPayload(
  source: CliBridgeAgent,
  rawPayload: string,
  env: Record<string, string | undefined> = process.env,
): CliHookMessage {
  let data: Record<string, unknown> = {};
  try {
    const parsed = JSON.parse(rawPayload);
    if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) data = parsed as Record<string, unknown>;
  } catch {
    data = {};
  }

  // Both supported agents default to PermissionRequest when the payload omits an
  // event name; `source` is statically claude|codex, so there is no other case.
  const hookEventName = String(data.hook_event_name ?? data.event_type ?? 'PermissionRequest');
  const normalized = normalizeEventName(hookEventName);
  const typeByEvent: Record<string, CliHookMessageType> = {
    PermissionRequest: 'permission_request',
    PostToolUse: 'post_tool_use',
    TaskComplete: 'task_complete',
    TaskCompleteFailure: 'task_complete',
  };
  const type: CliHookMessageType = typeByEvent[normalized ?? ''] ?? 'pre_tool_use';
  const toolInput = data.tool_input ?? data.toolInput ?? data.metadata ?? data.properties ?? {};

  return {
    type,
    source,
    sessionId: String(data.session_id ?? data.sessionId ?? ''),
    cwd: String(data.cwd ?? ''),
    toolName: String(data.tool_name ?? data.toolName ?? data.permission ?? ''),
    toolInput: toolInput && typeof toolInput === 'object' && !Array.isArray(toolInput) ? toolInput as Record<string, unknown> : {},
    hookEventName,
    stopHookActive: data.stop_hook_active === true || data.stopHookActive === true,
    permissionMode: typeof data.permission_mode === 'string' ? data.permission_mode : typeof data.permissionMode === 'string' ? data.permissionMode : undefined,
    permissionSuggestions: Array.isArray(data.permission_suggestions) ? data.permission_suggestions : Array.isArray(data.permissionSuggestions) ? data.permissionSuggestions : undefined,
    taskStatus: normalized === 'TaskCompleteFailure' ? 'failed' : normalized === 'TaskComplete' ? 'completed' : undefined,
    summary: type === 'task_complete' ? stringifySummaryValue(data) : undefined,
    bridgeOwned: env.FEISHU_CODEX_BRIDGE === '1',
    rawPayloadBytes: Buffer.byteLength(rawPayload, 'utf8'),
  };
}

export interface CliAskUserQuestion {
  questions: CliQuestionItem[];
}

/** Parse an AskUserQuestion / ask_user_question tool_input into 1-4 validated
 *  questions. Single- and multi-select are both accepted (the old single-question,
 *  single-select-only restriction is gone — the card now renders a multi-question
 *  form). Returns undefined (→ fall back to the local terminal) if the shape is
 *  unrecognized or any one question is malformed, so we never half-render garbage. */
export function extractAskUserQuestion(toolInput: Record<string, unknown>): CliAskUserQuestion | undefined {
  const rawQuestions = toolInput.questions;
  if (!Array.isArray(rawQuestions) || rawQuestions.length < 1 || rawQuestions.length > 4) return undefined;
  const questions = rawQuestions.flatMap((raw): CliQuestionItem[] => {
    if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return [];
    const obj = raw as Record<string, unknown>;
    const question = typeof obj.question === 'string' ? obj.question.trim() : '';
    if (!question) return [];
    const rawOptions = obj.options;
    if (!Array.isArray(rawOptions) || rawOptions.length < 2) return [];
    const options = rawOptions.flatMap((option) => {
      if (!option || typeof option !== 'object' || Array.isArray(option)) return [];
      const o = option as Record<string, unknown>;
      if (typeof o.label !== 'string' || !o.label.trim()) return [];
      return [{
        label: o.label.trim(),
        description: typeof o.description === 'string' && o.description.trim() ? o.description.trim() : undefined,
        preview: typeof o.preview === 'string' && o.preview.trim() ? o.preview.trim() : undefined,
      }];
    });
    if (options.length !== rawOptions.length) return [];
    return [{
      question,
      header: typeof obj.header === 'string' && obj.header.trim() ? obj.header.trim() : undefined,
      multiSelect: obj.multiSelect === true,
      options,
    }];
  });
  // Any one question failing validation drops the whole ask → local fallback.
  if (questions.length !== rawQuestions.length) return undefined;
  return { questions };
}

File: internal/clibridge/parser.go
package clibridge

// parser.go —— hook payload 解析（对齐 TS cli-bridge/parser.ts）。
// 把 agent hook 推来的原始 JSON（stdin）归一为 CliHookMessage；
// ExtractAskUserQuestion 把 AskUserQuestion tool_input 解析为 1-4 个校验过的问题。

import (
	"encoding/json"
	"strconv"
	"strings"
)

func normalizeEventName(eventName string) string {
	switch eventName {
	case "PermissionRequest", "permission_request", "permission.asked", "permission_requested":
		return "PermissionRequest"
	case "PreToolUse", "pre_tool_use":
		return "PreToolUse"
	case "PostToolUse", "post_tool_use":
		return "PostToolUse"
	case "Stop", "stop", "SubagentStop", "subagent_stop", "session.idle", "session_idle":
		return "TaskComplete"
	case "StopFailure", "stop_failure", "session.error", "session_error":
		return "TaskCompleteFailure"
	}
	return eventName
}

func stringifySummaryValue(value any) string {
	switch v := value.(type) {
	case nil:
		return ""
	case string:
		return strings.TrimSpace(v)
	case bool, float64, int:
		return toString(v)
	case []any:
		parts := make([]string, 0, len(v))
		for _, e := range v {
			if s := stringifySummaryValue(e); s != "" {
				parts = append(parts, s)
			}
		}
		return strings.TrimSpace(strings.Join(parts, "\n"))
	case map[string]any:
		for _, key := range []string{
			"last_assistant_message", "lastAssistantMessage", "assistant_message",
			"assistantMessage", "assistant", "final", "completion", "answer",
			"response", "output", "result", "content", "text", "message", "summary", "error",
		} {
			if s := stringifySummaryValue(v[key]); s != "" {
				return s
			}
		}
	}
	return ""
}

func toString(v any) string {
	switch x := v.(type) {
	case string:
		return x
	case bool:
		if x {
			return "true"
		}
		return "false"
	case float64:
		// 整数不加点。
		if x == float64(int64(x)) {
			return strconv.FormatInt(int64(x), 10)
		}
		return strconv.FormatFloat(x, 'f', -1, 64)
	case int:
		return strconv.FormatInt(int64(x), 10)
	}
	return ""
}

// ParseHookPayload 解析 hook 原始 JSON 为 CliHookMessage。
// env 携 bridgeOwned 标记（FEISHU_CODEX_BRIDGE=1）；传 nil 时视为无。
func ParseHookPayload(source CliBridgeAgent, rawPayload string, env map[string]string) CliHookMessage {
	var data map[string]any
	if raw := strings.TrimSpace(rawPayload); raw != "" {
		if err := json.Unmarshal([]byte(raw), &data); err != nil || data == nil {
			data = map[string]any{}
		}
	}
	if data == nil {
		data = map[string]any{}
	}

	hookEventName := stringOr(data["hook_event_name"], data["event_type"])
	if hookEventName == "" {
		hookEventName = "PermissionRequest"
	}
	normalized := normalizeEventName(hookEventName)

	typeByEvent := map[string]CliHookMessageType{
		"PermissionRequest": MsgTypePermissionRequest,
		"PostToolUse":       MsgTypePostToolUse,
		"TaskComplete":      MsgTypeTaskComplete,
		"TaskCompleteFailure": MsgTypeTaskComplete,
	}
	msgType := typeByEvent[normalized]
	if msgType == "" {
		msgType = MsgTypePreToolUse
	}

	toolInput := data["tool_input"]
	if toolInput == nil {
		toolInput = data["toolInput"]
	}
	if toolInput == nil {
		toolInput = data["metadata"]
	}
	if toolInput == nil {
		toolInput = data["properties"]
	}
	toolInputMap, _ := toolInput.(map[string]any)
	if toolInputMap == nil {
		toolInputMap = map[string]any{}
	}

	bridgeOwned := env["FEISHU_CODEX_BRIDGE"] == "1"

	return CliHookMessage{
		Type:       msgType,
		Source:     source,
		SessionID:  stringOr(data["session_id"], data["sessionId"]),
		Cwd:        stringOr(data["cwd"]),
		ToolName:   stringOr(data["tool_name"], data["toolName"], data["permission"]),
		ToolInput:  toolInputMap,
		HookEventName: hookEventName,
		StopHookActive: data["stop_hook_active"] == true || data["stopHookActive"] == true,
		PermissionMode: stringOr(data["permission_mode"], data["permissionMode"]),
		PermissionSuggestions: asAnySlice(data["permission_suggestions"], data["permissionSuggestions"]),
		TaskStatus: taskStatusOf(normalized),
		Summary:    func() string {
			if msgType == MsgTypeTaskComplete {
				return stringifySummaryValue(data)
			}
			return ""
		}(),
		BridgeOwned: bridgeOwned,
		RawPayloadBytes: len(rawPayload),
	}
}

// ExtractAskUserQuestion 把 AskUserQuestion/ask_user_question tool_input 解析为
// 1-4 个校验过的问题。任一问题非法 → 返回 nil（调用方回退本地终端）。
func ExtractAskUserQuestion(toolInput map[string]any) *struct {
	Questions []CliQuestionItem `json:"questions"`
} {
	rawQuestions, ok := toolInput["questions"].([]any)
	if !ok || len(rawQuestions) < 1 || len(rawQuestions) > 4 {
		return nil
	}
	questions := make([]CliQuestionItem, 0, len(rawQuestions))
	for _, raw := range rawQuestions {
		obj, ok := raw.(map[string]any)
		if !ok {
			return nil
		}
		question := strings.TrimSpace(stringOr(obj["question"]))
		if question == "" {
			return nil
		}
		rawOptions, ok := obj["options"].([]any)
		if !ok || len(rawOptions) < 2 {
			return nil
		}
		opts := make([]CliQuestionOption, 0, len(rawOptions))
		for _, o := range rawOptions {
			om, ok := o.(map[string]any)
			if !ok {
				return nil
			}
			label := strings.TrimSpace(stringOr(om["label"]))
			if label == "" {
				return nil
			}
			opts = append(opts, CliQuestionOption{
				Label:       label,
				Description: stringOrT(om["description"]),
				Preview:     stringOrT(om["preview"]),
			})
		}
		if len(opts) != len(rawOptions) {
			return nil
		}
		questions = append(questions, CliQuestionItem{
			Question:    question,
			Header:      stringOrT(obj["header"]),
			MultiSelect: obj["multiSelect"] == true,
			Options:     opts,
		})
	}
	if len(questions) != len(rawQuestions) {
		return nil
	}
	return &struct {
		Questions []CliQuestionItem `json:"questions"`
	}{Questions: questions}
}

// ── 小工具 ───────────────────────────────────────────────────

func stringOr(vals ...any) string {
	for _, v := range vals {
		if s, ok := v.(string); ok && s != "" {
			return s
		}
	}
	return ""
}

func stringOrT(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func asAnySlice(vals ...any) []any {
	for _, v := range vals {
		if s, ok := v.([]any); ok {
			return s
		}
	}
	return nil
}

func taskStatusOf(normalized string) string {
	switch normalized {
	case "TaskCompleteFailure":
		return "failed"
	case "TaskComplete":
		return "completed"
	}
	return ""
}
